client/batch.go

```diff
@@ -0,0 +1,426 @@
+// encoding.BinaryMarshaler. value can be any key type or a proto.Message.
+func (b *Batch) CPut(key, value, expValue interface{}) *Batch {
+	b = b.ensureNotNil()
+
+	k, err := marshalKey(key)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	v, err := marshalValue(value)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	ev, err := marshalValue(expValue)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	b.calls = append(b.calls, ConditionalPut(proto.Key(k), v, ev))
+	b.initResult(1, 1, nil)
+	return b
+}
+
+// Inc increments the integer value at key. If the key does not exist it will
+// be created with an initial value of 0 which will then be incremented. If the
+// key exists but was set using Put or CPut an error will be returned.
+//
+// A new result will be appended to the batch which will contain a single row
+// and Result.Err will indicate success or failure.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) Inc(key interface{}, value int64) *Batch {
+	b = b.ensureNotNil()
+
+	k, err := marshalKey(key)
+	if err != nil {
+		b.initResult(0, 1, err)
+		return b
+	}
+	b.calls = append(b.calls, Increment(proto.Key(k), value))
+	b.initResult(1, 1, nil)
+	return b
+}
+
+// Scan retrieves the rows between begin (inclusive) and end (exclusive).
+//
+// A new result will be appended to the batch which will contain up to maxRows
+// rows and Result.Err will indicate success or failure.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) Scan(s, e interface{}, maxRows int64) *Batch {
+	begin, err := marshalKey(s)
+	if err != nil {
+		b.initResult(0, 0, err)
+		return b
+	}
+	end, err := marshalKey(e)
+	if err != nil {
+		b.initResult(0, 0, err)
+		return b
+	}
+	b.calls = append(b.calls, Scan(proto.Key(begin), proto.Key(end), maxRows))
+	b.initResult(1, 0, nil)
+	return b
+}
+
+// Del deletes one or more keys.
+//
+// A new result will be appended to the batch and each key will have a
+// corresponding row in the returned Result.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) Del(keys ...interface{}) *Batch {
+	b = b.ensureNotNil()
+
+	var calls []Call
+	for _, key := range keys {
+		k, err := marshalKey(key)
+		if err != nil {
+			b.initResult(0, len(keys), err)
+			return b
+		}
+		calls = append(calls, Delete(proto.Key(k)))
+	}
+	b.calls = append(b.calls, calls...)
+	b.initResult(len(calls), len(calls), nil)
+	return b
+}
+
+// DelRange deletes the rows between begin (inclusive) and end (exclusive).
+//
+// A new result will be appended to the batch which will contain 0 rows and
+// Result.Err will indicate success or failure.
+//
+// key can be either a byte slice, a string, a fmt.Stringer or an
+// encoding.BinaryMarshaler.
+func (b *Batch) DelRange(s, e interface{}) *Batch {
```

Comment:
I'm not sure what the point of *returning* a pointer is. In general it's a good practice for methods to *either* mutate self or return a value, but not both (`append()` being a prominent exception in go). Returning self here allows for trivial cases to be written as one-liners, but I'm not sure that's interesting in the real world (except for the prevalence of one-method "batches" to be combined with txn.Commit. I'm not happy with the current ways to express that idiom but I don't have any better ideas at the moment).

If Batch methods took self by value instead of pointer, then you would be required to use the return value from each chained call; a multi-line use case would look like

    b := client.B()
    b = b.Put(a, b)
    b = b.Put(c, d)

It would also be expensive: the batch struct includes a fixed-size buffer to reduce allocation overhead and we don't want to copy that by value (although I'm not sure how important that optimization is)